File: src/utils/scoring.ts
import type { Board, BoardSize, Player, Position, ScoreResult } from "../types/game.js";
import { cloneBoard, getNeighbors, posKey } from "./board.js";

export interface RemoveDeadStonesResult {
  board: Board;
  deadBlack: number;
  deadWhite: number;
}

/**
 * Removes every stone whose posKey is in `deadStones` from the board, returning the
 * cleaned board plus how many stones of each color were removed — those counts become
 * captures for the opponent when the final score is calculated.
 */
export function removeDeadStones(board: Board, deadStones: Set<string>): RemoveDeadStonesResult {
  const cleaned = cloneBoard(board);
  let deadBlack = 0;
  let deadWhite = 0;

  for (const key of deadStones) {
    const [row, col] = key.split(",").map(Number);
    const stone = cleaned[row][col];
    if (stone === "black") deadBlack++;
    else if (stone === "white") deadWhite++;
    if (stone !== null) cleaned[row][col] = null;
  }

  return { board: cleaned, deadBlack, deadWhite };
}

interface EmptyRegion {
  region: Position[];
  borderColors: Set<Player>;
}

function floodFillEmptyRegion(board: Board, start: Position, size: BoardSize, visited: Set<string>): EmptyRegion {
  const region: Position[] = [];
  const borderColors = new Set<Player>();
  const stack: Position[] = [start];
  visited.add(posKey(start));

  while (stack.length > 0) {
    const pos = stack.pop()!;
    region.push(pos);

    for (const neighbor of getNeighbors(pos, size)) {
      const stone = board[neighbor.row][neighbor.col];
      const key = posKey(neighbor);

      if (stone === null) {
        if (!visited.has(key)) {
          visited.add(key);
          stack.push(neighbor);
        }
      } else {
        borderColors.add(stone);
      }
    }
  }

  return { region, borderColors };
}

/**
 * Simple area-style scoring: stones on the board + surrounded territory + captures,
 * plus komi (a fixed compensation added to White's score to offset Black's first-move
 * advantage — see types/game.ts::KOMI_OPTIONS). Territory is only awarded to a connected
 * region of empty points bordered by exactly one color; regions touching both colors
 * (or no stones at all) are neutral.
 */
export function calculateScore(
  board: Board,
  size: BoardSize,
  blackCaptures: number,
  whiteCaptures: number,
  komi: number
): ScoreResult {
  let blackStones = 0;
  let whiteStones = 0;
  const blackTerritoryPoints: Position[] = [];
  const whiteTerritoryPoints: Position[] = [];
  const visited = new Set<string>();

  for (let row = 0; row < size; row++) {
    for (let col = 0; col < size; col++) {
      const stone = board[row][col];

      if (stone === "black") {
        blackStones++;
        continue;
      }
      if (stone === "white") {
        whiteStones++;
        continue;
      }

      const key = posKey({ row, col });
      if (visited.has(key)) continue;

      const { region, borderColors } = floodFillEmptyRegion(board, { row, col }, size, visited);
      if (borderColors.size === 1) {
        const [owner] = borderColors;
        if (owner === "black") blackTerritoryPoints.push(...region);
        else whiteTerritoryPoints.push(...region);
      }
    }
  }

  const blackTerritory = blackTerritoryPoints.length;
  const whiteTerritory = whiteTerritoryPoints.length;
  const blackScore = blackStones + blackTerritory + blackCaptures;
  const whiteScore = whiteStones + whiteTerritory + whiteCaptures + komi;

  let winner: Player | "draw";
  if (blackScore > whiteScore) winner = "black";
  else if (whiteScore > blackScore) winner = "white";
  else winner = "draw";

  return {
    blackStones,
    whiteStones,
    blackTerritory,
    whiteTerritory,
    blackTerritoryPoints,
    whiteTerritoryPoints,
    blackCaptures,
    whiteCaptures,
    komi,
    blackScore,
    whiteScore,
    winner,
  };
}

File: ai-service/src/scoring.py
"""Area scoring, mirroring src/utils/scoring.ts::calculateScore (stones + surrounded
territory + captures + komi) -- the exact same rule the app uses to decide a winner.

The one deliberate difference: the TS version takes a pre-cleaned board (dead stones
already removed by removeDeadStones, based on what the players marked in the UI).
MCTS has no one to mark dead stones -- a simulated line reaching two passes gets scored
on the raw board as-is (every stone counted alive). This is a real limitation on
messy/unsettled positions, but it's the same well-defined algorithm the app already
uses, not a new or different rule; see mcts.py's module docstring for why this is an
acceptable approximation for search rather than a scoring bug.
"""

from dataclasses import dataclass
from typing import List, Set, Tuple

from src.adapters.game_adapter import Board, Player


def _neighbors(row: int, col: int, size: int):
    if row > 0:
        yield (row - 1, col)
    if row < size - 1:
        yield (row + 1, col)
    if col > 0:
        yield (row, col - 1)
    if col < size - 1:
        yield (row, col + 1)


@dataclass
class ScoreResult:
    black_stones: int
    white_stones: int
    black_territory: int
    white_territory: int
    black_score: float
    white_score: float
    winner: str  # "black" | "white" | "draw"


def calculate_score(
    board: Board, size: int, black_captures: int, white_captures: int, komi: float
) -> ScoreResult:
    black_stones = 0
    white_stones = 0
    black_territory = 0
    white_territory = 0
    visited: Set[Tuple[int, int]] = set()

    for row in range(size):
        for col in range(size):
            stone = board[row][col]
            if stone == "black":
                black_stones += 1
                continue
            if stone == "white":
                white_stones += 1
                continue

            if (row, col) in visited:
                continue

            # Flood-fill this empty region; territory belongs to a color only if the
            # whole connected region borders exactly that one color (never both).
            region: List[Tuple[int, int]] = []
            border_colors: Set[Player] = set()
            stack = [(row, col)]
            visited.add((row, col))
            while stack:
                r, c = stack.pop()
                region.append((r, c))
                for nr, nc in _neighbors(r, c, size):
                    neighbor = board[nr][nc]
                    if neighbor is None:
                        if (nr, nc) not in visited:
                            visited.add((nr, nc))
                            stack.append((nr, nc))
                    else:
                        border_colors.add(neighbor)

            if len(border_colors) == 1:
                (owner,) = border_colors
                if owner == "black":
                    black_territory += len(region)
                else:
                    white_territory += len(region)

    black_score = black_stones + black_territory + black_captures
    white_score = white_stones + white_territory + white_captures + komi

    winner = "black" if black_score > white_score else "white" if white_score > black_score else "draw"

    return ScoreResult(
        black_stones=black_stones,
        white_stones=white_stones,
        black_territory=black_territory,
        white_territory=white_territory,
        black_score=black_score,
        white_score=white_score,
        winner=winner,
    )
